This exercise will require you to pull some data from the Qunadl API. Qaundl is currently the most widely used aggregator of financial market data.

As a first step, you will need to register a free account on the http://www.quandl.com website.

After you register, you will be provided with a unique API key, that you should store:

In [19]:
# Store the API key as a string - according to PEP8, constants are always named in all upper case
API_KEY = 'agrZBA8v5xQkAcznmeu8'

Qaundl has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Quandl API instructions here: https://docs.quandl.com/docs/time-series

While there is a dedicated Python package for connecting to the Quandl API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/ ).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [2]:
# First, import the relevant modules
import requests


In [78]:
# Now, call the Quandl API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
#https://www.quandl.com/api/v3/datasets/{database_code}/{dataset_code}/data.{return_format}
url = "https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json?start_date=2017-11-16&end_date=2017-11-17&api_key=agrZBA8v5xQkAcznmeu8"
 
params = dict(api_key=API_KEY)
res = requests.get(url)



{'dataset': {'id': 10095370,
  'dataset_code': 'AFX_X',
  'database_code': 'FSE',
  'name': 'Carl Zeiss Meditec (AFX_X)',
  'description': 'Stock Prices for Carl Zeiss Meditec (2020-08-31) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
  'refreshed_at': '2020-11-13T22:16:17.522Z',
  'newest_available_date': '2020-11-06',
  'oldest_available_date': '2000-06-07',
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'frequency': 'daily',
  'type': 'Time Series',
  'premium': False,
  'limit': None,
  'transform': None,
  'column_index': None,
  'start_date': '2017-11-16',
  'end_date': '2017-11-17',
  'data': [['2017-11-17',
    47.03,
    47.15,
    46.8,
    46.84,
    None,
    54107.0,
    2540820.0,
    None,
    None,
    None],
   ['2017-11-16',
    47.09,
    47.23,
    46.55,
    47.03,
    None

In [80]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
json_api = res.json()
json_api

{'dataset': {'id': 10095370,
  'dataset_code': 'AFX_X',
  'database_code': 'FSE',
  'name': 'Carl Zeiss Meditec (AFX_X)',
  'description': 'Stock Prices for Carl Zeiss Meditec (2020-08-31) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
  'refreshed_at': '2020-11-13T22:16:17.522Z',
  'newest_available_date': '2020-11-06',
  'oldest_available_date': '2000-06-07',
  'column_names': ['Date',
   'Open',
   'High',
   'Low',
   'Close',
   'Change',
   'Traded Volume',
   'Turnover',
   'Last Price of the Day',
   'Daily Traded Units',
   'Daily Turnover'],
  'frequency': 'daily',
  'type': 'Time Series',
  'premium': False,
  'limit': None,
  'transform': None,
  'column_index': None,
  'start_date': '2017-11-16',
  'end_date': '2017-11-17',
  'data': [['2017-11-17',
    47.03,
    47.15,
    46.8,
    46.84,
    None,
    54107.0,
    2540820.0,
    None,
    None,
    None],
   ['2017-11-16',
    47.09,
    47.23,
    46.55,
    47.03,
    None

The json is made up many keys under the key dataset. We are interested in the column_names and data, the data is a list of lists however.

In [92]:
#to create a dictionary with column name as keys and data as values
#since the data in the json file is a list of lists so it needs to be unpacked and ziped first before creating a dict
#of keys and values
afx_dict = {i:list(j) for i, j in zip(json_api['dataset']['column_names'],zip(*json_api['dataset']['data']))}
afx_dict

{'Date': ['2017-11-17', '2017-11-16'],
 'Open': [47.03, 47.09],
 'High': [47.15, 47.23],
 'Low': [46.8, 46.55],
 'Close': [46.84, 47.03],
 'Change': [None, None],
 'Traded Volume': [54107.0, 89373.0],
 'Turnover': [2540820.0, 4195732.0],
 'Last Price of the Day': [None, None],
 'Daily Traded Units': [None, None],
 'Daily Turnover': [None, None]}

In [90]:
print(list(zip(*json_api['dataset']['data'])))

[('2017-11-17', '2017-11-16'), (47.03, 47.09), (47.15, 47.23), (46.8, 46.55), (46.84, 47.03), (None, None), (54107.0, 89373.0), (2540820.0, 4195732.0), (None, None), (None, None), (None, None)]


These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

### 1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017

In [99]:
#Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017
url_2017 = "https://www.quandl.com/api/v3/datasets/FSE/AFX_X.json"
params = dict(start_date='2017-1-1',end_date='2017-12-31',api_key=API_KEY)
res_2017 =requests.get(url,params = params)
res_2017


<Response [200]>

### 2. Convert the returned JSON object into a Python dictionary

In [110]:
#Convert the returned JSON object into a Python dictionary
json_2017= res_2017.json()
dict_afx_2017 = {i:list(j) for i, j in zip(json_2017['dataset']['column_names'],zip(*json_2017['dataset']['data']))}

In [112]:
#keys of dict
dict_afx_2017.keys()

dict_keys(['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'])

### 3. Calculate what the highest and lowest opening prices were for the stock in this period


In [192]:
#create list of opening prices

lst_open_prices = dict_afx_2017['Open']

In [193]:
#check all elements of a l are the same type, float
all(isinstance(elem, float) for elem in lst_open_prices)

False

Not all elements in the list are floats 

In [202]:
#function that counts the types in a list, takes in a list and returns a dict with the types
def list_types(lst):
    d= {}
    for elem in lst:
        typ = type(elem)
        if typ in d:
            d[typ] += 1
        else:
            d[typ] = 1
    return d


In [205]:
#checking types of lst_open_prices
lst_types = list_types(lst_open_prices)
lst_types

{float: 252, NoneType: 3}

There are three NoneType, which means empty values

In [214]:
#get rid of nones and test 
open_prices = [item for item in lst_open_prices if item is not None]
list_types(open_prices)

{float: 252}

In [216]:
#Calculate the highest opening prices
max(open_prices)

53.11

Highest opening price = $53.11

In [217]:
#4. Calculate the lowest opening price
min(open_prices)

34.0

The lowest opening price = $34

### 4.What was the largest change in any one day (based on High and Low price)?

In [218]:
dict_afx_2017.keys()

dict_keys(['Date', 'Open', 'High', 'Low', 'Close', 'Change', 'Traded Volume', 'Turnover', 'Last Price of the Day', 'Daily Traded Units', 'Daily Turnover'])

In [237]:
#check types of daily lows and highs
high = dict_afx_2017['High']
low = dict_afx_2017['Low']
print(list_types(high))
print(list_types(low))

{<class 'float'>: 255}
{<class 'float'>: 255}


In [240]:
#largest change in one day
diff = [abs(h-l) for h,l in zip(high,low)]
max_chg = max(diff)
max_chg

2.8100000000000023

Largest change in any one day = $2.81

### 5.What was the largest change between any two days (based on Closing Price)?

In [243]:
#function that takes in a list and returns largest difference between any two elements
def max_chg(lst):
    chg = 0
    for elem1 in lst:
        for elem2 in lst:
            diff = abs(elem1 - elem2)
            if diff > chg:
                chg = diff
    return chg

In [244]:
max_chg(dict_afx_2017['Close'])

19.03

largest change between any two days = $19.03

### 6.What was the average daily trading volume during this year?

In [262]:
#function to calculate average of a list
 #takes in a list and returns the list's average
def average(lst):
    avg = sum(lst) / len(lst)
    return avg

In [265]:
#average traded volume
average(dict_afx_2017['Traded Volume'])

89124.33725490196

Average daily trading volume during this year = 89124.34 AFX stocks

### 6.(Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [321]:
def get_median(lst):
    sort_lst = sorted(lst)
    half_len = len(lst) // 2
    if half_len%2 == 0:
        med = (sort_lst[half_len - 1] + sort_lst[half_len])//2
    else:
        med = sort_lst[half_len]
    return sort_lst

In [275]:
255 // 2

127

In [323]:
l=[2,1, 4, 3]
half_len = len(l) // 2
len(l)//2

2

In [324]:
get_median(l)

[1, 2, 3, 4]

In [326]:
l[(len(l)//2)-1]

1